copy/copy.go

```diff
@@ -1041,22 +1084,45 @@ type diffIDResult struct {
 	err    error
 }
 
+type imageSourceSeekableProxy struct {
+	source   types.ImageSourceSeekable
+	progress chan int64
+}
+
+func (s imageSourceSeekableProxy) GetBlobAt(ctx context.Context, bInfo types.BlobInfo, chunks []types.ImageSourceChunk) (io.ReadCloser, string, error) {
+	rc, contentType, err := s.source.GetBlobAt(ctx, bInfo, chunks)
+	if err == nil {
+		total := int64(0)
+		for _, c := range chunks {
+			total += int64(c.Length)
+		}
+		s.progress <- total
+	}
+	return rc, contentType, err
+}
+
 // copyLayer copies a layer with srcInfo (with known Digest and Annotations and possibly known Size) in src to dest, perhaps compressing it if canCompress,
 // and returns a complete blobInfo of the copied layer, and a value for LayerDiffIDs if diffIDIsNeeded
 func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, toEncrypt bool, pool *mpb.Progress) (types.BlobInfo, digest.Digest, error) {
 	cachedDiffID := ic.c.blobInfoCache.UncompressedDigest(srcInfo.Digest) // May be ""
 	// Diffs are needed if we are encrypting an image or trying to decrypt an image
 	diffIDIsNeeded := ic.diffIDsAreNeeded && cachedDiffID == "" || toEncrypt || (isOciEncrypted(srcInfo.MediaType) && ic.c.ociDecryptConfig != nil)
 
+	srcAlgo := ""
+	if srcInfo.CompressionAlgorithm != nil {
```

Comment:
`srcInfo.Compression*` is supposed to never be set at this point; they really don’t belong to `BlobInfo` at all, it’s an output of `copyBlobFromStream` to be consumed by `UpdatedImage`.